File: typescript/search/LargestNumberAtLeastTwiceOfOthers.ts
/**
 * LeetCode 747. Largest Number At Least Twice of Others
 * https://leetcode.com/problems/largest-number-at-least-twice-of-others/
 */
function dominantIndex(nums: number[]): number {
    let largestNumIdx: number = -1;
    let largestNum: number = Number.MIN_SAFE_INTEGER;
    let secondLargestNum: number = Number.MIN_SAFE_INTEGER;
    for (let i: number = 0; i < nums.length; i++) {
        if (nums[i] > largestNum) {
            secondLargestNum = largestNum;
            largestNum = nums[i];
            largestNumIdx = i;
        } else if (nums[i] > secondLargestNum && nums[i] <= largestNum) {
            secondLargestNum = nums[i];
        }
    }
    return largestNum < secondLargestNum * 2 ? -1 : largestNumIdx;
};

// EOF

File: go/search/LargestNumberAtLeastTwiceOfOthers.go
package main

import (
	"math"
)

/**
 * LeetCode 747. Largest Number At Least Twice of Others
 * https://leetcode.com/problems/largest-number-at-least-twice-of-others/
 */
func dominantIndex(nums []int) int {
	if nums == nil || len(nums) < 2 {
		return 0
	}
	largestNumIdx := -1
	largestNum := int(math.MinInt32)
	secondLargestNum := int(math.MinInt32)
	for i := 0; i < len(nums); i++ {
		if nums[i] > largestNum {
			secondLargestNum = largestNum
			largestNum = nums[i]
			largestNumIdx = i
		} else if nums[i] > secondLargestNum && nums[i] <= largestNum {
			secondLargestNum = nums[i]
		}
	}
	if largestNum < secondLargestNum*2 {
		return -1
	}
	return largestNumIdx
}

func main() {
	// ...
}

// EOF
